### ISSUE TITLE
How to get a model with good generalization ability

### ISSUE THREAD

**Issue by changshuowang** [VimsLab/EllipsoidQuery#4]
Hello,author .
        Your work is very meaningful, but I have encountered some problems.I got a training accuracy of 0.926661 from your code, but on the test set, its effect is always between 0.92-0.93, as shown in the figure below. But the accuracy of your open source ('cls/model_cls_ssn_iter_70917_acc_0.925446.pth') on the test set is basically maintained above 0.93. Excuse me, why is the generalization ability of the model I trained so poor? Is there a problem where I am? How do I get a good training model on the test set next? Hope to get your reply, thank you!
![image](https://user-images.githubusercontent.com/49676110/108692456-977af180-7537-11eb-83f3-c94bb0158c75.png)

**Comment by sheshap** [comment]
@changshuowang Thanks for reporting the observations you made.
RS-CNN uses 300 evaluations to choose the highest accuracy.
The pre-trained model accuracy achieves above 0.93 (as confirmed by you) and the highest being 0.935 (We used 100 evaluations to choose the best).

Please let me know if you are using the same software configurations as recommended in the readme file.
Ubuntu 18.04
Python 3.5 (recommend Anaconda3)
Pytorch 0.4.1
CMake 3.10.2
CUDA 10.0 + cuDNN 7
Cudatoolkit V10.0.130

I will also check for the issue you mentioned for fresh training and update you.

**Comment by changshuowang** [comment]
@sheshap Thank you brother for your reply. My environment configuration is as follows:
Ubuntu 18.04.5
python 3.6.12
pytorch 0.4.1
CMake 3.10.2
CUDA 10.2 + cuDNN 7.6.5
Cudatoolkit 9.0

I have trained for many times, and the best accuracy rate can reach 93.35%, but in most cases it is lower than 93%. Is it enough to train for the best model results?

In addition, I have two other questions: 
1. Brother, have you used your method on multiple scales? Is the accuracy rate higher than that of a single scale?
2. I see that many methods of extracting point clouds are to select the best training model through the test set, and then vote and test on the test set. Is this reasonable?

I am honored to discuss these issues with you and look forward to your reply to these issues.
Thank you.

**Comment by sheshap** [comment]
@changshuowang Thank you for bringing up these questions and also for letting me know that you got 93.35%. 

1. Using the ellipsoid method for multiple scales is complex because it is harder to search for appropriate ellipsoid dimensions per scale and per set abstractions.  On a single scale, only 2-3 ellipsoids are enough whereas for multiple-scale it could go up to 9-12 different ellipsoids. Each ellipsoid has 3 axes. So, it will be a lot of numbers. Of course, need more GPUs to train. 

2. Yes, you are right about many recent methods, they select the best training model, vote, and then test. The trend is forcing everyone to adopt that pipeline. It is surely not the best practice because the best results could be due to randomness while training. 

I have observed a change in the accuracy due to randomness from recent works.
[https://github.com/j1a0m0e4sNTU/3dgcn/issues/5](https://github.com/j1a0m0e4sNTU/3dgcn/issues/5)
[https://github.com/liruihui/PointAugment/issues/8](https://github.com/liruihui/PointAugment/issues/8)
[https://github.com/hkust-vgd/scanobjectnn/issues/22](https://github.com/hkust-vgd/scanobjectnn/issues/22)

Also, accuracies might increase or decrease due to different implementations (using pytorch).
[https://github.com/yanx27/Pointnet_Pointnet2_pytorch](https://github.com/yanx27/Pointnet_Pointnet2_pytorch)

Please feel free to reach me at shiv093@gmail.com. Happy to discuss.
